File: core/src/test/kotlin/org/stellar/anchor/util/PropertyUtilTest.java
package org.stellar.anchor.util;

import static org.junit.jupiter.api.Assertions.*;
import static org.stellar.anchor.util.PropertyUtil.set;

import java.util.Optional;
import lombok.Data;
import org.junit.jupiter.api.Test;

class PropertyUtilTest {
  @Data
  public static class A {
    String value;
    B b;
  }

  @Data
  public static class B {
    String value;
    C c;
  }

  @Data
  public static class C {
    String value;
  }

  @Test
  void testSet() throws ReflectiveOperationException {
    A a = new A();
    assertEquals(Optional.empty(), PropertyUtil.get(a, "value"));
    set(a, "value", "value-a");
    assertEquals("value-a", PropertyUtil.get(a, "value").orElse(null));

    assertEquals(Optional.empty(), PropertyUtil.get(a, "b.value"));
    set(a, "b.value", "value-b");
    assertEquals("value-b", PropertyUtil.get(a, "b.value").orElse(null));

    assertEquals(Optional.empty(), PropertyUtil.get(a, "b.c.value"));
    set(a, "b.c.value", "value-c");
    assertEquals("value-c", PropertyUtil.get(a, "b.c.value").orElse(null));
  }

  @Test
  void testNotExists() throws ReflectiveOperationException {
    A a = new A();
    assertEquals(Optional.empty(), PropertyUtil.get(a, "value"));
    assertEquals(Optional.empty(), PropertyUtil.get(a, "b"));
    assertEquals(Optional.empty(), PropertyUtil.get(a, "b.c"));
    assertEquals(Optional.empty(), PropertyUtil.get(a, "not_found"));
  }
}

File: core/src/test/kotlin/org/stellar/anchor/util/PropertyUtilTest.kt
package org.stellar.anchor.util

import java.util.*
import lombok.Data
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

internal class PropertyUtilTest {
  @Data
  class A {
    var value: String? = null
    var b: B? = null
  }

  @Data
  class B {
    var value: String? = null
    var c: C? = null
  }

  @Data
  class C {
    var value: String? = null
  }

  @Test
  fun `test set property`() {
    val a = A()
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "value"))
    PropertyUtil.set(a, "value", "value-a")
    assertEquals("value-a", PropertyUtil.get(a, "value").orElse(null))
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "b.value"))
    PropertyUtil.set(a, "b.value", "value-b")
    assertEquals("value-b", PropertyUtil.get(a, "b.value").orElse(null))
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "b.c.value"))
    PropertyUtil.set(a, "b.c.value", "value-c")
    assertEquals("value-c", PropertyUtil.get(a, "b.c.value").orElse(null))
  }

  @Test
  fun `test get non-existent property returns empty`() {
    val a = A()
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "value"))
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "b"))
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "b.c"))
    assertEquals(Optional.empty<Any>(), PropertyUtil.get(a, "not_found"))
  }
}
